### ISSUE TITLE
String cache management

### ISSUE THREAD

**Issue by tamasfe** [pola-rs/polars#2857]
I've been using polars for a week or so and it's been a great experience so far! However I struggle(d) with understanding how string caches for categorical data are managed and why they are handled the way it is done currently.

This issue is written from the POV of a new user without too much knowledge of the internals so my observations might or might not align with the reality, bear with me.

### Struggles with the (Python) API

My use-case and thought process was the following:

1. enable a string cache
1. build multiple frames (and series) that share similar or the same categorical data
1. do multiple operations on them that involve multiple lazy frames and potentially some `collect` operations in between
1. clear the string cache after everything is done

Now, intuitively for me this would be achievable in the following ways:

```py
pl.Config.set_global_string_cache()
do_whatever()
pl.Config.unset_global_string_cache()
```

or similarly

```py
with pl.StringCache():
    do_whatever()
```

The main issue is that lazy frames also manage (**and implicitly clear**) the very same global string cache with each execution, which can be opted out from, but then as far as I see, no string cache is used at all (correct me on this one).

### Potential race conditions

As far as I see the same global string cache is accessed by multiple operations (casting data and execution for example), and since it's possible to execute multiple dataframe operations on different threads, I believe it's possible that right after casting a series, another execution on another thread clears the cache before the next operation.

#### Explicit string cache management

Allow the creation of string caches and have the ability to pass them to operations as needed, this puts more control in the user's hands who most likely knows how the data is used.

#### Lazy frames should never implicitly clear the cache

This is probably the most confusing feature right now, as it seems to make it impossible to reuse the same string cache across multiple operations.

#### Explicit global cache

The global cache could be opt-in like it is now for users who don't want granular management and would still like to benefit from the advantages of a cache.

I believe the following should be distinct operations and none of them should be done implicitly:

- enabling the global cache
- disabling the global cache
- clearing the global cache

As global caching is opt-in, the user of the API should be made aware of the caveats (potential race conditions) and the need to clear the cache explicitly, this way one can make sure that the cache is intact between operations even during multiple executions on different threads.

#### Python API

With the above points implemented, I imagine the Python API would look something like the examples below. Similar approaches can be used in Rust as well. I also didn't take backward compatibility into account yet, but it can be done later if this proposal is accepted.

For global caches:

```py
pl.GlobalStringCache.enable()
operations_with_cache()

# lazy frames can still opt-out individually with the current API
pl.DataFrame().lazy().collect(string_cache=False)

pl.GlobalStringCache.disable()

# global cache is disabled for this operation
operations_without_cache()

# use the same global cache
pl.GlobalStringCache.enable()
operations_with_cache()
pl.GlobalStringCache.disable()

# we're done with the entire data set
pl.GlobalStringCache.clear()
```

```py
with pl.GlobalStringCache():
    operations_with_cache()

operations_without_cache()

# also clear the cache when the guard goes out of scope.
with pl.GlobalStringCache(clear=True):
    operations_with_cache()
```

For individual caches:

```py
cache = pl.StringCache()

pl.DataFrame().lazy().select([]).collect(string_cache=cache)

# caches could be implicitly destroyed/cleared in destructors as well
cache.clear()

# with a guard that clears the string cache at the end
with pl.StringCache() as sc:
    bar = (
        pl.DataFrame()
        .select(pl.col("foo").cast(pl.Categorical, string_cache=sc))
        .lazy()
        .collect(string_cache=sc)
    )
```

**Comment by ritchie46** [comment]
I agree with you. It was a mistake to create a string cache per lazy computation graph as they are global and not bound to that graph. 

It is a complexity that should be removed and I agree the user should decide when to use and stop using the global string cache.

**Comment by ritchie46** [comment]
closed by #2916
